### PR TITLE
refactor(api): move module commands out of the hardware controller

### LINKED ISSUE #7799: Hardware API: Refactor module commands from the hardware controller
# Overview
To make the hardware test api easier, and more extensible, we should first start cutting down the code in `api.py` to make it easier for exposing the hardware controller in separable parts.

# Acceptance criteria
- All module commands should now live in its own separate file
- Push down module handling to the hardware controller to deduplicate code

### PR DESCRIPTION
**Pull request by Laura-Danielle**
# Overview
There were a couple of goals with this refactor. First, we wanted to pull out all of the functionality around registering modules/finding registered modules. Second, I wanted to clean up the watch function for aionotify (maybe we can find a better library for this use-case that is cross-platform compatible, can get a bit annoying for typing purposes). Third, I wanted to make the initial discovery part of modules a little cleaner. Closes #7799.

# Changelog
- Clean up watch functionality in the controllers
- Created a new class to hold module registration/deregistration/finding control

# Future Considerations
1. I would love to combine all of the system 'watches' together/give full control to the controller for this. In a follow-up PR, we should refactor how the door file change is watched.
2. I feel like the execution manager should also be handled by the controller, but I'm open to other opinions on keeping it in the api file.

# Review requests
Please thoroughly test on a robot/review the code. Let me know if things don't make sense or you'd like me to change how something is implemented. Also let me know if naming of things is confusing.

# Risk assessment

High! Pulled out all the module code, deleted/refactored how modules are initially scanned. We should:
1. Run protocols with multiple modules
2. Check modules of the same type functionality still works
3. Plug/unplug modules a bunch of times, check that the app updates in a reasonable time.
4. Control modules in the app, check that that still works.

### CODE REVIEW

**Review by sfoster1 — APPROVED** [review]
Big fan, looks great!

**Review by ahiuchingau — APPROVED** [review]
Makes sense that the modules should be handled by the controller instead of the api. code LGTM, but needs to be tested on the robot still

File: api/tests/opentrons/hardware_control/test_modules.py [hunk 1]
```diff
@@ -68,7 +68,7 @@ async def test_filtering_modules():
         'magdeck', 'thermocycler']
     api = await hardware_control.API.build_hardware_simulator(
                         attached_modules=mods)
-    await asyncio.sleep(0.5)
+    await asyncio.sleep(0.05)
```
Comment (ahiuchingau) [inline api/tests/opentrons/hardware_control/test_modules.py:71]: Just curious: why did you need to change this?

File: api/src/opentrons/hardware_control/api.py [hunk 2]
```diff
@@ -74,6 +75,8 @@ def __init__(self,
         self._config = config
         self._backend = backend
         self._loop = loop
+        # TODO (lc 05-12-2021) give responsibility of the the
+        # execution manager to the hardware controller in a follow-up.
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/api.py]: It's not clear to me what this TODO statement means

File: api/src/opentrons/hardware_control/module_control.py [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+
+    async def handle_module_appearance(self, event):
+        """ Only called upon availability of aionotify. Check that
+        the file system has changed and either remove or add modules
+        depending on the result.
+
+        :param event: Aionotify event. Note that since aionotify is
+        only supported on a linux system, we cannot properly type
+        this parameter.
+        """
```
Comment (amitlissack) [inline api/src/opentrons/hardware_control/module_control.py:201]: Not necessarily for this PR. 

But relates to your comment about `aionotify`.

It would be nice to isolate the `aionotify` knowledge in one place. In this PR the _event_watcher is owned by `Controller`, right? The `Controller` just delegates events to the `AttachedModuleControl` class, right? 
It would be nice if the `event` was some data structure of our own creation with no connection to `aionotify`.

**Review by amitlissack — APPROVED** [review]
I love seeing code in this package split up. 

This is a nice step in the right direction.  I see us making significant changes in module discovery down the road. I would love to be able to hook our software emulations into the hardware controller.

File: api/tests/opentrons/hardware_control/test_modules.py (hunk 1 above)
Comment (Laura-Danielle) [reply api/tests/opentrons/hardware_control/test_modules.py:71]: I didn't need to sleep for as long as I was in this test. Matching the sleep time to all of the other tests, to speed up test time.

File: api/src/opentrons/hardware_control/module_control.py (hunk 3 above)
Comment (Laura-Danielle) [reply api/src/opentrons/hardware_control/module_control.py:201]: I'll change that for this PR. It would be nice to potentially find an alternative to aionotify as well.

File: api/src/opentrons/hardware_control/api.py (hunk 2 above)
Comment (Laura-Danielle) [reply api/src/opentrons/hardware_control/api.py]: I should remove it. I'm not sure whether we should actually follow-through with this 'TODO' or not. Right now, the execution manager is being used in `api.py` as a gate to "readiness" for performing an action. To me, it's a very similar action to the motion lock in `api.py` but is closer to protocol execution rather than actual motion on the robot. In my opinion, we should ideally be controlling motion at a lower level than the `api.py` file because we do not care about those sorts of things in simulation.

Similarly, (correct me if I'm wrong in this thinking) we do not care about the protocol 'state' during simulation either (i.e. whether a pause or delay happens in simulation, we don't care and generally skip over it).

Comment (amitlissack) [reply api/src/opentrons/hardware_control/api.py]: I agree with you. It would be nice to separate protocol play/pause/cancel from HW play/pause.

**Review by nusrat813 — APPROVED** [review]
Tested - LGTM
